src/graph/network/socket_communicator.h

```diff
@@ -32,118 +33,147 @@ struct Addr {
 };
 
 /*!
- * \brief Network Sender for DGL distributed training.
+ * \brief SocketSender for DGL distributed training.
  *
- * Sender is an abstract class that defines a set of APIs for sending 
- * binary data over network. It can be implemented by different underlying 
- * networking libraries such TCP socket and ZMQ. One Sender can connect to 
- * multiple receivers, and it can send data to specified receiver via receiver's ID.
+ * SocketSender is the communicator implemented by tcp socket.
  */
 class SocketSender : public Sender {
  public:
   /*!
-   * \brief Add receiver address and it's ID to the namebook
-   * \param ip receviver's IP address
-   * \param port receiver's port
+   * \brief Sender constructor
+   * \param queue_size size of message queue 
+   */
+  explicit SocketSender(int64_t queue_size) : Sender(queue_size) {}
+
+  /*!
+   * \brief Add receiver's address and ID to the sender's namebook
+   * \param addr Networking address, e.g., 'socket://127.0.0.1:50091', 'mpi://0'
    * \param id receiver's ID
    */
-  void AddReceiver(const char* ip, int port, int recv_id);
+  void AddReceiver(const char* addr, int recv_id);
 
   /*!
    * \brief Connect with all the Receivers
-   * \return True for sucess and False for fail
+   * \return True for success and False for fail
    */
   bool Connect();
 
   /*!
-   * \brief Send data to specified Receiver
+   * \brief Send data to specified Receiver. Actually pushing data into message queue.
    * \param data data buffer for sending
-   * \param size data size for sending
+   * \param size size of data for sending
    * \param recv_id receiver's ID
-   * \return bytes we sent
+   * \return bytes of data
    *   > 0 : bytes we sent
    *   - 1 : error
+   *
+   * Note that, DO NOT delete the data pointer in Send() because we use zero-copy here. 
+   * The Send() API is a non-blocking API, which will return from function immediantely 
+   * if the message queue is not full, and it will be blocked when message queue is full.
+   *
+   * The Send() API is thread-safe, but we DO NOT guarantee the order of message 
+   * in multi-threading sending.
```

Comment:
For my implementation,  if we set deallocator to nullptr, the deallocator will do nothing on the data, because sometimes users don't want to delete their message in communicator.